registry/servicediscovery/client.go

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) The OpenTofu Authors
+// SPDX-License-Identifier: MPL-2.0
+
+package servicediscovery
+
+import (
+	"context"
+	"net/http"
+
+	tofuhttp "github.com/opentofu/libregistry/registry/internal/http"
+)
+
+type Client interface {
+	ServiceDiscovery(ctx context.Context, request Request) (Response, error)
+}
+
+type ClientOpt func(cfg *config) error
+
+func NewClient(opts ...ClientOpt) (Client, error) {
+	cfg := config{}
+	for _, opt := range opts {
+		if err := opt(&cfg); err != nil {
+			return nil, err
+		}
+	}
+	cfg.applyDefaults()
+	return &client{
+		cfg,
+	}, nil
+}
+
+func WithHTTPClient(client *http.Client) ClientOpt {
+	return func(cfg *config) error {
+		cfg.client = client
+		return nil
+	}
+}
+
+func WithEndpoint(endpoint string) ClientOpt {
```

Comment:
Having this be `WithHostname` (or `ForHostname`?) and taking a [`svchost.Hostname`](https://pkg.go.dev/github.com/hashicorp/terraform-svchost#Hostname) might make this package tessellate better with how surrounding packages work.

For example, [`tfaddr.Provider`](https://pkg.go.dev/github.com/opentofu/registry-address#Provider) and [`tfaddr.ModulePackage`](https://pkg.go.dev/github.com/opentofu/registry-address#ModulePackage) from the "registry-address" module both provide the hostname portion of the source address as a `svchost.Hostname`, and so someone using both of these libraries together could just pass the hostname over here without needing first transform it into a full base URL. (That's also, incidentally, how [the corresponding code in OpenTofu itself](https://github.com/opentofu/opentofu/blob/7fddff6ba4f6e2338361fcd9b33587eae6d09709/internal/getproviders/registry_source.go#L112-L113)  is treating it.)

Relatedly, [RFC 8615](https://www.rfc-editor.org/rfc/rfc8615) requires that the `/.well-known/` segment always appear at the start of the path, so it would be non-compliant for the discovery document to be placed anywhere other than the root of the authority implied by the `svchost.Hostname` value. It seems reasonable to expose that situation as an alternative for contrived situations like unit tests, but I think it's preferable for the main entry point to force a spec-compliant implementation, especially when that also allows some consistency with the other library.

Reply:
There is quite a stack of assumptions built on the idea that service providers are identified by so-called "friendly hostnames", including their use in the source address syntax for both providers and registry-published modules, so it isn't really clear to me how we would get there from where we currently are, but I imagine that's a much bigger discussion than this PR.

It seems strange to me for this client to prematurely offer such flexibility before we've designed how that flexibility might manifest in the rest of the system. I would expect this library to implement the protocols as they are currently defined first, and then evolve along with the rest of the system if we choose to change the protocols later.